### ISSUE TITLE
Feature Request: Add optional oldText field to TextEdit

### ISSUE THREAD

**Issue by ShaneDelmore** [microsoft/language-server-protocol#474]
I am happy to make a proposal for this feature if others agree it is a welcome addition.

TextEdit currently has range and newText fields, I would like to add oldText to allow clients to optionally verify the server is making a valid recommendation.  My example use case is on rename, if I make a rename request to rename symbol Foo to Bar and the server returns the TextEdit
```
{
  "Range": "SomeRange",
  "NewText": "Bar",
  "OldText": " Fo" //"Foo" expected
}
```
On the client side I can see that the server is out of date and resync the document, then request the rename again.

**Comment by dbaeumer** [comment]
Actually the idea here is that the rename returns a workspace edit that uses `documentChanges` instead of `changes`. `TextDocumentEdit` the type of `documentChanges` have a version text document identifier which contains a version number the change is valid for. The server is allowed to set the version number to null if the document is not open hence the server has not version information and the truth lives on disk.

Doesn't this work for you?

**Comment by ShaneDelmore** [comment]
Darn, I was away from github for a bit and this issue timed out.  @dbaeumer The reason oldText would be nice for my use case is that I use edit distance to try to create a valid rename even if the file has been changed slightly.  This prevents needing to do a full recompile on every change, but means my results may occasionally be incorrect, which oldText would allow the clien tot detect.

**Comment by dbaeumer** [comment]
I am reopening the issue. Since the field will be optional I am pretty sure most servers will not set it.

**Comment by ShaneDelmore** [comment]
Is the next step that I should submit a PR with the changes then?

**Comment by dbaeumer** [comment]
Actually I am not a supporter of this so it needs some broader discussion. As said the field can only be optional and then I suspect most servers will not set it so you will not be able to rely on it and profit from it. May be we should think about whether we add some mechanisms to request the edits on the new document state. Or you could keep some older states around to actually fill in the old text.

**Comment by ShaneDelmore** [comment]
The core problem for me is that I have a centralized server that is very useful to clients, but expected to not be 100% accurate.  If there were a way I could return a document version the client could verify the version was accurate or not, but that conflicts with the idea of LSP having a client->server supplied specific incrementing version.  A sha256 of the file for example would be enough to know if the versions matched or not.  Is there a place this might fit into the spec?  On large projects there is considerable value in recalculating my lsp indexes, but that raises the question of having a way to know when they are out of date. Additional ideas on how to best accomplish this are welcomed.

**Comment by rcjsuen** [comment]
> On the client side I can see that the server is out of date and resync the document, then request the rename again.

How is the client supposed to ask the server to "resync" the document? With a `textDocument/didChange` request that includes the file in its entirety?

> If there were a way I could return a document version the client could verify the version was accurate or not, but that conflicts with the idea of LSP having a client->server supplied specific incrementing version. A sha256 of the file for example would be enough to know if the versions matched or not. Is there a place this might fit into the spec?

Could you rephrase this? Are you suggesting a new `textDocument/serverVersion` or whatever request that a client would to the server to see if the server's version number (or SHA hash or whatever) matches the one that the client is currently at?

**Comment by ShaneDelmore** [comment]
To restate (including my assumptions in case they are incorrect)

1.   I will start up my lsp client and connect to my server and they will have disk contents out of sync
2.  I would be fairly expensive to resync them and mostly useless.
3.  I do not have shared disk between the server and client
4.  The server can still provide useful information to the client, but the client may need to do some filtering to see if the information applies.

Whether through my original oldText suggestion, or more closely aligned with @dbaeumer suggestion to use DocumentVersion, through a version identifier that does not require co-ordination like a sha instead of an incrementing integer, I need a way on the client to know if the server recommendation is applicable.

Regarding the “resync”, a textDocument/didChange request that includes the file in it’s entirety would work well.

**Comment by rcjsuen** [comment]
Thank you for clarifying, @ShaneDelmore. Your comment about not having the same disk shared between the client and the server is certainly not the average use case. That clears things up a little.

Regarding your original example, should `oldText` be `"Foo"` instead of `" Fo"`? Or maybe I'm looking at the example in the wrong way...?

Consider the client's file and the server's file below.
```
client:
foo
server:
foo
```
You insert a `f` at `{ line: 0, character: 0 }` on the client.
```
client:
ffoo
server:
foo
```
But you then ask for a rename at the `position` on `{ line: 0, character: 2 }` and `newName` of `bar`. In the client's case, this is the cursor position between `f` and `o` whereas in the server's case, the cursor's position is between the first `o` and the second `o`. The server confirms that the position of the text caret is over a valid symbol that can be renamed and sends the information back.
```
{
  range: {
    start: { line: 0, character: 0},
    end: { line: 0, character: 3 }
  },
  newText: "bar",
  oldText: "foo" // this is the text in that text range on the server's side
}
```
The client receives this and sees the range. It compares the text locally and the text at that range is `ffo` which is not equal to `foo` and rejects the edit.

Is the above understanding correct?

**Comment by ShaneDelmore** [comment]
Exactly!

**Comment by ShaneDelmore** [comment]
I know this isn’t a common use case, but in larger code bases it really provides a nice benefit to have a remotely precomputed index, even if it is supplemented by a local server, this allows the local server to only compute what has changed from the remote resulting in considerable efficiency improvements.
